File: apps/backend/internal/model/marketplace.go
package model

import "time"

// Listing represents a marketplace listing (lot-based, SPL token)
type Listing struct {
	ID           string    `json:"id" db:"id"`
	ProjectID    string    `json:"projectId" db:"project_id"`
	SellerID     string    `json:"sellerId" db:"seller_id"`
	TokenID      string    `json:"tokenId" db:"token_id"`
	ScaledAmount int64     `json:"scaledAmount" db:"scaled_amount"`
	Price        float64   `json:"price" db:"price"`
	PaymentToken string    `json:"paymentToken" db:"payment_token"`
	Active       bool      `json:"active" db:"active"`
	CreatedAt    time.Time `json:"createdAt" db:"created_at"`
	UpdatedAt    time.Time `json:"updatedAt" db:"updated_at"`
}

// DisplayAmount returns the human-readable credit amount (scaledAmount / 1000)
func (l Listing) DisplayAmount() float64 {
	return float64(l.ScaledAmount) / 1000.0
}

// ListingWithDetails includes project and seller information
type ListingWithDetails struct {
	Listing
	ProjectTitle        string  `json:"projectTitle" db:"project_title"`
	ProjectDescription  string  `json:"projectDescription" db:"project_description"`
	ProjectImageURL     string  `json:"projectImageUrl" db:"project_image_url"`
	SellerEmail         string  `json:"sellerEmail" db:"seller_email"`
	MintAddress         *string `json:"mintAddress" db:"mint_address"`
	TokenSymbol         *string `json:"tokenSymbol" db:"token_symbol"`
	SellerWalletAddress *string `json:"sellerWalletAddress" db:"seller_wallet_address"`
}

File: packages/zod/src/marketplace.ts
import { z } from 'zod'

// ============================================================================
// Marketplace Listing Schemas
// ============================================================================

export const CreateListingSchema = z.object({
    projectId: z.string().uuid('Invalid project ID'),
    tokenId: z.number().int().nonnegative('Invalid token ID'),
    amount: z.number().positive('Amount must be greater than 0'),
    priceEth: z.number().positive('Price must be greater than 0')
})

export const BuyListingSchema = z.object({
    listingId: z.string().uuid('Invalid listing ID'),
    sourceLotId: z.number().int().nonnegative().optional()
})

export const CancelListingSchema = z.object({
    listingId: z.string().uuid('Invalid listing ID')
})

export const GetListingSchema = z.object({
    listingId: z.string().uuid('Invalid listing ID')
})

export const ListMarketplaceSchema = z.object({
    page: z.number().int().positive().optional().default(1),
    limit: z.number().int().positive().max(100).optional().default(20),
    projectId: z.string().uuid().optional(),
    active: z.boolean().optional()
})

export const ListingSchema = z.object({
    id: z.string().uuid(),
    projectId: z.string().uuid(),
    sellerId: z.string(),
    tokenId: z.number().int().nonnegative(),
    scaledAmount: z.number().int().nonnegative(),
    priceEth: z.number(),
    active: z.boolean(),
    createdAt: z.string(),
    updatedAt: z.string()
})

export const ListingWithDetailsSchema = ListingSchema.extend({
    projectTitle: z.string(),
    sellerEmail: z.string().email(),
    tokenAddress: z.string(),
    tokenSymbol: z.string()
})

// ============================================================================
// Token Minting Schemas
// ============================================================================

export const MintTokenSchema = z.object({
    projectId: z.string().uuid('Invalid project ID'),
    amount: z.number().positive('Amount must be greater than 0'),
    toAddress: z.string().regex(/^0x[a-fA-F0-9]{40}$/, 'Invalid Ethereum address')
})

export const DeployProjectTokenSchema = z.object({
    projectId: z.string().uuid('Invalid project ID')
})

// ============================================================================
// Type Exports
// ============================================================================

export type CreateListing = z.infer<typeof CreateListingSchema>
export type BuyListing = z.infer<typeof BuyListingSchema>
export type CancelListing = z.infer<typeof CancelListingSchema>
export type GetListing = z.infer<typeof GetListingSchema>
export type ListMarketplace = z.infer<typeof ListMarketplaceSchema>
export type Listing = z.infer<typeof ListingSchema>
export type ListingWithDetails = z.infer<typeof ListingWithDetailsSchema>
export type MintToken = z.infer<typeof MintTokenSchema>
export type DeployProjectToken = z.infer<typeof DeployProjectTokenSchema>
